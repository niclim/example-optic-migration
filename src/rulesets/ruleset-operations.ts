import { OperationRule, RuleError, Ruleset } from "@useoptic/rulesets-base";

export const preventOperationRemoval = new OperationRule({
  name: "prevent operation removal",
  rule: (operationAssertions) => {
    operationAssertions.removed(() => {
      throw new RuleError({
        message: "cannot remove an operation. This is a breaking change.",
      });
    });
  },
});

export const OperationRuleset = new Ruleset({
  name: "operations",
  rules: [preventOperationRemoval],
});
