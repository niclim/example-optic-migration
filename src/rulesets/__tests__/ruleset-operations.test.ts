import { test, expect, describe } from "@jest/globals";
import { OpenAPIV3 } from "@useoptic/openapi-utilities";
import { TestHelpers } from "@useoptic/rulesets-base";
import { OperationRuleset } from "../ruleset-operations";

describe("ruleset operation", () => {
  test("operation removal", async () => {
    const beforeJson: OpenAPIV3.Document = {
      ...TestHelpers.createEmptySpec(),
      paths: {
        "/api/users": {
          get: {
            responses: {
              "200": {
                description: "",
              },
            },
          },
        },
      },
    };
    const afterJson: OpenAPIV3.Document = {
      ...TestHelpers.createEmptySpec(),
      paths: {},
    };
    const results = await TestHelpers.runRulesWithInputs(
      [OperationRuleset],
      beforeJson,
      afterJson
    );
    expect(results.length > 0).toBe(true);
    expect(results).toMatchSnapshot();
    expect(results.some((result) => !result.passed)).toBe(true);
  });
});
