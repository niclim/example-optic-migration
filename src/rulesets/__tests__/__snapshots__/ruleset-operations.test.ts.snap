// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ruleset operation operation removal 1`] = `
[
  {
    "change": {
      "changeType": "removed",
      "location": {
        "conceptualLocation": {
          "method": "get",
          "path": "/api/users",
        },
        "conceptualPath": [
          "operations",
          "/api/users",
          "get",
        ],
        "jsonPath": "/paths/~1api~1users/get",
        "kind": "operation",
      },
      "removed": {
        "before": {
          "method": "get",
          "pathPattern": "/api/users",
        },
      },
    },
    "condition": undefined,
    "docsLink": undefined,
    "error": "cannot remove an operation. This is a breaking change.",
    "exempted": false,
    "expected": undefined,
    "isMust": true,
    "isShould": false,
    "name": "prevent operation removal",
    "passed": false,
    "received": undefined,
    "type": "removed",
    "where": "GET /api/users",
  },
]
`;
